### PR TITLE
bug(number-field): update number-field value on pressing "enter"

### PR DESCRIPTION
**Pull request by TarunAdobe**
## Description

If you type in a number larger than the max attribute and press "return", it should change the number to the max.
i.e. The value inside the number-field should update on pressing "return" however right now it only happens when the input is blurred out.

## Related issue(s)

- #3000

## Motivation and context

If you type in a number larger than the max attribute and press "return", it doesn't always change the number. However, the number is always changed when the input is blurred. In addition, the number is never changed when format-options is used, and the number is only changed the first time you press "return" even without format-options.

## How has this been tested?

-   [ ] _Test case 1_
    1. Go here
    2. Do this
-   [ ] _Test case 2_
    1. Go here
    2. Do this

## Types of changes

-   [x] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)
-   [ ] Chore (minor updates related to the tooling or maintenance of the repository, does not impact compiled assets)

## Checklist

-   [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
-   [x] My code follows the code style of this project.
-   [x] If my change required a change to the documentation, I have updated the documentation in this pull request.
-   [x] I have read the **[CONTRIBUTING](<(https://github.com/adobe/spectrum-web-components/blob/main/CONTRIBUTING.md)>)** document.
-   [ ] I have added tests to cover my changes.
-   [x] All new and existing tests passed.
-   [x] I have reviewed at the Accessibility Practices for this feature, see: [Aria Practices](https://www.w3.org/TR/wai-aria-practices/)

## Best practices

This repository uses conventional commit syntax for each commit message; note that the GitHub UI does not use this by default so be cautious when accepting suggested changes. Avoid the "Update branch" button on the pull request and opt instead for rebasing your branch against `main`.

### CODE REVIEW

**Comment by blunteshwar** [comment]
@TarunAdobe some VRTs are failing. Can you look at them

**Review by Rajdeepc — DISMISSED** [review]

File: packages/number-field/src/NumberField.ts [hunk 1]
```diff
@@ -340,6 +340,13 @@ export class NumberField extends TextfieldBase {
                     new Event('change', { bubbles: true, composed: true })
                 );
                 break;
+            case 'Enter':
+                event.preventDefault();
```
Comment (Westbrook) [inline packages/number-field/src/NumberField.ts:344]: What is the default that you're preventing here?

We haven't gotten to Form Association for these Custom Elements, but in the future, were this associated to a form, there would be some expectation that `Enter` would submit the form by default.

File: packages/number-field/src/NumberField.ts [hunk 2]
```diff
@@ -340,6 +340,13 @@ export class NumberField extends TextfieldBase {
...
+                this.inputElement.value = this.formattedValue;
+                this.dispatchEvent(
+                    new Event('change', { bubbles: true, composed: true })
+                );
```
Comment (Westbrook) [inline packages/number-field/src/NumberField.ts:348]: This looks good are there any tests that would be beneficial to add in support of this going forward?

File: packages/number-field/src/NumberField.ts (hunk 1 above)
Comment (TarunAdobe) [reply packages/number-field/src/NumberField.ts]: good point! I have removed the `event.preventDefault()` and I've updated the few tests with the current behaviour expectations

File: packages/number-field/src/NumberField.ts [hunk 3]
```diff
@@ -160,6 +160,12 @@ export class NumberField extends TextfieldBase {
         }
         const oldValue = this._value;
         this._value = value;
+        if (!this.managedInput && this.lastCommitedValue != this.value) {
```
Comment (Westbrook) [inline packages/number-field/src/NumberField.ts:163]: `!=` allows coercion, where at `!==` does not. I'm not sure we'd ever be in a situation for coercion to occur here, but if it's possible to use instead, let's normalize on `!==` when possible.

File: packages/number-field/src/NumberField.ts [hunk 4]
```diff
@@ -400,7 +405,11 @@ export class NumberField extends TextfieldBase {
             }
         }
         this.value = value;
```
Comment (Westbrook) [inline packages/number-field/src/NumberField.ts:407]: This line will trigger the setter. Will the following lines _never_ accidentally dispatch a second `change` event?

File: packages/number-field/test/number-field.test.ts [hunk 5]
```diff
@@ -566,7 +566,7 @@ describe('NumberField', () => {
                 ],
             });
             expect(inputSpy.callCount).to.equal(4);
-            expect(changeSpy.callCount).to.equal(1);
+            expect(changeSpy.callCount).to.equal(0); // the actual value hasn't changed so changespy won't be called
```
Comment (Westbrook) [inline packages/number-field/test/number-field.test.ts:569]: Great point! The name of this test is no longer valid.

We may want both this test and the previous test, just make sure it doesn't end on the same value as it started...

Comment (Westbrook) [reply packages/number-field/test/number-field.test.ts:569]: Also...

```suggestion
            expect(changeSpy.callCount, 'value does not change from initial value so no "change" event is dispatched').to.equal(0);
```

The `expect` method accepts "comments" as the second argument.

File: packages/number-field/test/number-field.test.ts [hunk 6]
```diff
@@ -751,7 +751,7 @@ describe('NumberField', () => {
             await sendKeys({ press: 'Enter' });
             await elementUpdated(el);
             expect(lastInputValue, 'last input value').to.equal(10);
-            expect(lastChangeValue, 'last change value').to.equal(10);
+            expect(lastChangeValue, 'last change value').to.equal(0); // value wouldn't go beyond max, so it remains unchanged after pressing enter and thus no changeSpy called
```
Comment (Westbrook) [inline packages/number-field/test/number-field.test.ts:754]: A change and not a change likely benefit two versions of this, too.

File: packages/number-field/src/NumberField.ts (hunk 3 above)
Comment (Westbrook) [reply packages/number-field/src/NumberField.ts:163]: Also, will this dispatch a `change` event if the use use `el.value = 'new value'` (but a valid value) from the outside?

Comment (Westbrook) [reply packages/number-field/src/NumberField.ts:163]: If this doesn't do that, we should add a test to confirm for the future. If it does do that, we should add a test to confirm that the alternate approach that you take to get there, too.

**Review by Westbrook — COMMENTED** [review]
Really like what you've settled on here! A couple of extra cases to confirm and tests to duplicate before we can merge this, but it's really starting to shape up.

File: packages/number-field/src/NumberField.ts (hunk 4 above)
Comment (TarunAdobe) [reply packages/number-field/src/NumberField.ts:407]: no because in the setter we are updated the lastCommittedValue which will cause the check on the line 409 to fail making sure there is not second `change` event

File: packages/number-field/src/NumberField.ts (hunk 3 above)
Comment (TarunAdobe) [reply packages/number-field/src/NumberField.ts]: thnx for noticing this! updated it🥂
